<div style="border:solid Chocolate 2px; padding: 40px">

<b>Привет!👋</b>

Меня зовут Огурлиев Анзор, я буду ревьюером твоего проекта. Сразу хочу предложить тебе общаться на "ты", если тебе это неудобно, скажи мне об этом. Нам предстоит пройти с тобой через тернистый и от того интересный путь анализа данных! Можешь не переживать, я всегда помогу тебе, если у тебя возникнут вопросы)

По ходу твоей работы я буду оставлять свои комментарии, по выполненным тобой шагам. Попрошу тебя не удалять их, так мне будет проще отслеживать твой прогресс в случае повторной проверки. 
    
Очень важно вести диалог и обсуждать те пункты, которые у тебя могли не получиться, поэтому, ты можешь оставлять свои комментарии, и мы обязательно во всем разберемся! Самое главное - выделяй свои комментарии удобным для тебя способом (цвет, рамка), но так, чтобы они отчетлитво отличались от моих комментариев. Так мы ни в чем не запутаемся)

Свои комментарии я буду выделять так: <font color='green'>зеленый</font>, <font color='gold'>желтый</font> и <font color='red'>красный</font> цвет, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Этот комментарий я буду оставлять, если твое решение оптимально для данного шага.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> Это не означает, что ты совершил ошибку. Такими комментариями я буду отмечать места, где ты справился с задачей, но её можно решить ещё лучше. Также в таких комментариях я буду оставлять для тебя другие советы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Если решение, которое ты применил на определенном шаге, требует значительных изменений, то я буду оставлять такой комментарий. Тебе нужно будет подумать еще раз как правильным образом решить данный этап. Если ревью содержит такие комментарии, то я не смогу принять проект с первого раза.
</div>
    
P.S. В оглавлении комментария ты увидишь `V1` (`V1`, `V2`, `V3`) - это версия моего комментария, которая понадобится нам с тобой, чтобы не запутаться, если потребуется оставить тебе комментарии после первой проверки.

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Выясним уникальные значения количества детей
data['children'].value_counts()

0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

In [23]:
# Выясним влияет ли начилие детей на образования долга по кредиту и тоже самое без детей
with_children = data[data['children'] != 0]['children'].count()
with_children_debt = data[(data['children'] != 0) & (data['debt'] == 1)]['children'].count()
without_children = data[data['children'] == 0]['children'].count()
without_children_debt = data[(data['children'] == 0) & (data['debt'] == 1)]['children'].count()

conversion_with_children = with_children_debt / with_children
conversion_without_children = without_children_debt / without_children

print(f'Конверсия для заемщиков с детьми равна: {conversion_with_children:.3%}')
print(f'Конверсия для заемщиков без детей равна: {conversion_without_children:.3%}')

Конверсия для заемщиков с детьми равна: 9.240%
Конверсия для заемщиков без детей равна: 7.544%


In [24]:
# Выясним как влияет количество детей на образования долга по кредиту
with_children_1 = data[data['children'] == 1]['children'].count()
with_children_debt_1 = data[(data['children'] == 1) & (data['debt'] == 1)]['children'].count()
with_children_2 = data[data['children'] == 2]['children'].count()
with_children_debt_2 = data[(data['children'] == 2) & (data['debt'] == 1)]['children'].count()
with_children_3 = data[data['children'] == 3]['children'].count()
with_children_debt_3 = data[(data['children'] == 3) & (data['debt'] == 1)]['children'].count()
with_children_4 = data[data['children'] == 4]['children'].count()
with_children_debt_4 = data[(data['children'] == 4) & (data['debt'] == 1)]['children'].count()
with_children_5 = data[data['children'] == 5]['children'].count()
with_children_debt_5 = data[(data['children'] == 5) & (data['debt'] == 1)]['children'].count()

conversion_with_children_1 = with_children_debt_1 / with_children_1
conversion_with_children_2 = with_children_debt_2 / with_children_2
conversion_with_children_3 = with_children_debt_3 / with_children_3
conversion_with_children_4 = with_children_debt_4 / with_children_4
conversion_with_children_5 = with_children_debt_5 / with_children_5

print(f'Конверсия для заемщиков с одним ребенком равна: {conversion_with_children_1:.3%}')
print(f'Конверсия для заемщиков с двумя детьми равна: {conversion_with_children_2:.3%}')
print(f'Конверсия для заемщиков с тремя детьми равна: {conversion_with_children_3:.3%}')
print(f'Конверсия для заемщиков с четырьмя равна: {conversion_with_children_4:.3%}')
print(f'Конверсия для заемщиков с пятью детьми равна: {conversion_with_children_5:.3%}')

Конверсия для заемщиков с одним ребенком равна: 9.235%
Конверсия для заемщиков с двумя детьми равна: 9.454%
Конверсия для заемщиков с тремя детьми равна: 8.182%
Конверсия для заемщиков с четырьмя равна: 9.756%
Конверсия для заемщиков с пятью детьми равна: 0.000%


In [25]:
with_children_debt_1

444

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Нужно построить сводную таблицу с помощью метода `groupby()` или `pivot_table()`. В данной таблице должна быть вся необходимая информация для анализа. Нужно вывести такие столбцы: количество клиентов, количество просрочек, доля просрочек по группам.
</div>

In [26]:
data_pivot_children = data.pivot_table(index=['children'], values=['debt'], aggfunc=['sum', 'count'])
data_pivot_children.columns=['Кол-во просрочек', 'Количество заемщиков']
data_pivot_children['Доля просрочек'] = data_pivot_children['Кол-во просрочек'] / data_pivot_children['Количество заемщиков'] *100
data_pivot_children = data_pivot_children.sort_values(by = 'Доля просрочек')
data_pivot_children

Кол-во просрочек  Количество заемщиков  Доля просрочек
children                                                        
5                        0                     9        0.000000
0                     1063                 14091        7.543822
3                       27                   330        8.181818
1                      444                  4808        9.234609
2                      194                  2052        9.454191
4                        4                    41        9.756098

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Правильно построил сводную таблицу, молодец!
</div>

**Вывод:** 
1. Заемщики без детей берут кредиты активнее.
2. Заемщики с детьми оформляют кредиты менее активно, но при этом риск возникновения задолженности выше, чем у заемщиков без детей.
3. Количество детей не сильно влияет на изменение образования задолженности.

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Именно так, все верно.
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [27]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Выясним уникальные значения семейного положения
data['family_status'].value_counts()

женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64

In [28]:
# Выясним есть ли зависимость от семейного положения об разования задолженности
family_status_0 = data[data['family_status_id'] == 0]['family_status_id'].count()
family_status_debt_0 = data[(data['family_status_id'] == 0) & (data['debt'] == 1)]['family_status_id'].count()
family_status_1 = data[data['family_status_id'] == 1]['family_status_id'].count()
family_status_debt_1 = data[(data['family_status_id'] == 1) & (data['debt'] == 1)]['family_status_id'].count()
family_status_2 = data[data['family_status_id'] == 2]['family_status_id'].count()
family_status_debt_2 = data[(data['family_status_id'] == 2) & (data['debt'] == 1)]['family_status_id'].count()
family_status_3 = data[data['family_status_id'] == 3]['family_status_id'].count()
family_status_debt_3 = data[(data['family_status_id'] == 3) & (data['debt'] == 1)]['family_status_id'].count()
family_status_4 = data[data['family_status_id'] == 4]['family_status_id'].count()
family_status_debt_4 = data[(data['family_status_id'] == 4) & (data['debt'] == 1)]['family_status_id'].count()

conversion_family_status_0 = family_status_debt_0 / family_status_0
conversion_family_status_1 = family_status_debt_1 / family_status_1
conversion_family_status_2 = family_status_debt_2 / family_status_2
conversion_family_status_3 = family_status_debt_3 / family_status_3
conversion_family_status_4 = family_status_debt_4 / family_status_4

print(f'Конверсия для заемщиков в статусе женат/замужем равна: {conversion_family_status_0:.3%}')
print(f'Конверсия для заемщиков в статусе гражданский брак равна: {conversion_family_status_1:.3%}')
print(f'Конверсия для заемщиков в статусе вдовец/вдова равна: {conversion_family_status_2:.3%}')
print(f'Конверсия для заемщиков в статусе в разводе равна: {conversion_family_status_3:.3%}')
print(f'Конверсия для заемщиков в статусе не женат/не замужем равна: {conversion_family_status_4:.3%}')

Конверсия для заемщиков в статусе женат/замужем равна: 7.561%
Конверсия для заемщиков в статусе гражданский брак равна: 9.313%
Конверсия для заемщиков в статусе вдовец/вдова равна: 6.625%
Конверсия для заемщиков в статусе в разводе равна: 7.065%
Конверсия для заемщиков в статусе не женат/не замужем равна: 9.764%


<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
В этом и во всех последующих пунктах нужно построить сводные таблицы, как я описал выше.
</div>

In [29]:
data_pivot_family_status = data.pivot_table(index=['family_status'], values=['debt'], aggfunc=['sum', 'count'])
data_pivot_family_status.columns=['Кол-во просрочек', 'Количество заемщиков']
data_pivot_family_status['Доля просрочек'] = data_pivot_family_status['Кол-во просрочек'] / data_pivot_family_status['Количество заемщиков'] *100
data_pivot_family_status = data_pivot_family_status.sort_values(by = 'Доля просрочек')
data_pivot_family_status

Кол-во просрочек  Количество заемщиков  Доля просрочек
family_status                                                                
вдовец / вдова                       63                   951        6.624606
в разводе                            84                  1189        7.064760
женат / замужем                     927                 12261        7.560558
гражданский брак                    385                  4134        9.313014
Не женат / не замужем               273                  2796        9.763948

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Видно, что ты хорошо справляешься с методами построения сводных таблиц, так держать!
</div>

**Вывод:** 
1. Наменьший риск представляют заемщики со статусами вдовец/вдова и в разводе.
2. Наибольший риск представляют заемщики со статусами гражданский брак и не женат/не замужем.
3. Заемщики с наименьшим риском, люди более возрастные и подходящие к выбору осторожнее. А заемщики, которые представляют наиблльший риск, люди более молодые и менее осторожные. А гражданский брак не накладывает той ответственности, который несет официальный статут (женат/замужем).

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Все верно, клиенты, побывавшие в браке, реже допускают просрочки.
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [30]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Так как уровни дохода были раньше обозначены, то сразу перейдем к расчету зависимости

total_income_a = data[data['total_income_category'] == 'A']['total_income_category'].count()
total_income_debt_a = data[(data['total_income_category'] == 'A') & (data['debt'] == 1)]['total_income_category'].count()
total_income_b = data[data['total_income_category'] == 'B']['total_income_category'].count()
total_income_debt_b = data[(data['total_income_category'] == 'B') & (data['debt'] == 1)]['total_income_category'].count()
total_income_c = data[data['total_income_category'] == 'C']['total_income_category'].count()
total_income_debt_c = data[(data['total_income_category'] == 'C') & (data['debt'] == 1)]['total_income_category'].count()
total_income_d = data[data['total_income_category'] == 'D']['total_income_category'].count()
total_income_debt_d = data[(data['total_income_category'] == 'D') & (data['debt'] == 1)]['total_income_category'].count()
total_income_e = data[data['total_income_category'] == 'E']['total_income_category'].count()
total_income_debt_e = data[(data['total_income_category'] == 'E') & (data['debt'] == 1)]['total_income_category'].count()

conversion_total_income_a = total_income_debt_a / total_income_a
conversion_total_income_b = total_income_debt_b / total_income_b
conversion_total_income_c = total_income_debt_c / total_income_c
conversion_total_income_d = total_income_debt_d / total_income_d
conversion_total_income_e = total_income_debt_e / total_income_e


print(f'Конверсия для заемщиков с доходом 1000001 и выше равна: {conversion_total_income_a:.3%}')
print(f'Конверсия для заемщиков с доходом 200001–1000000 равна: {conversion_total_income_b:.3%}')
print(f'Конверсия для заемщиков с доходом 50001–200000 равна: {conversion_total_income_c:.3%}')
print(f'Конверсия для заемщиков с доходом 30001–50000 равна: {conversion_total_income_d:.3%}')
print(f'Конверсия для заемщиков с доходом 0–30000 равна: {conversion_total_income_e:.3%}')

Конверсия для заемщиков с доходом 1000001 и выше равна: 8.000%
Конверсия для заемщиков с доходом 200001–1000000 равна: 7.060%
Конверсия для заемщиков с доходом 50001–200000 равна: 8.498%
Конверсия для заемщиков с доходом 30001–50000 равна: 6.017%
Конверсия для заемщиков с доходом 0–30000 равна: 9.091%


**Вывод:** 
1. Из данных видно, что по образованию задолженности риск представляют заемщики среднего класса.
2. Неудивительно, что наибольший риск представляют заемщики с низким уровнем дохода.
3. Удивительно, что заемщики с высоким уровнем дохода имеют довольно высокий риск образования задолженности.

<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Обрати внимание на распределение клиентов по группам. Можно ли учитывать все группы при выявлении зависимости?
    
Подсказка: зависимость нужно определять только по 2 группам с достаточным количеством клиентов.
</div>

Привет, Анзор! 
Первый раз обращаюсь к тебе и сразу на "ты", так как ты просил на "ты".
Не буду тебе воду лить про всю "эту" ситуацию (СВО) и о том, что только как неделю назад попал в конечную точку.
Я лишь хочу извиниться, что так сильно задерживаю с обраткой. Система пустила дальше изучать и я еще раз изивиняюсь за время. 



По поводу определения только двух групп.
Согласен с тобой по большому счету, но думаю так. Если нам дали разделение/классификацию (при условии, что действительно нам дали, а не сами ее придумали) по доходам, то заказчик хотел бы видеть, пусть даже и малое количество, но по всей классификации.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Так заказчик же не имеет достоверной уверенности в том, что все группы содержат достаточное количество наблюдений, в реальной жизни таких ситуаций будет очень мало.
</div>

In [31]:
data_pivot_total_income = data.pivot_table(index=['total_income_category'], values=['debt'], aggfunc=['sum', 'count'])
data_pivot_total_income.columns=['Кол-во просрочек', 'Количество заемщиков']
data_pivot_total_income['Доля просрочек'] = data_pivot_total_income['Кол-во просрочек'] / data_pivot_total_income['Количество заемщиков'] *100
data_pivot_total_income = data_pivot_total_income.sort_values(by = 'Доля просрочек')
data_pivot_total_income.index = ['от 30 001 до 50 000', 'от 200 001 до 1 000 000', 'от 1 000 001 и выше', 'от 50 001 до 200 000', 'от 0 до 30 000']
data_pivot_total_income

Кол-во просрочек  Количество заемщиков  \
от 30 001 до 50 000                    21                   349   
от 200 001 до 1 000 000               354                  5014   
от 1 000 001 и выше                     2                    25   
от 50 001 до 200 000                 1353                 15921   
от 0 до 30 000                          2                    22   

                         Доля просрочек  
от 30 001 до 50 000            6.017192  
от 200 001 до 1 000 000        7.060231  
от 1 000 001 и выше            8.000000  
от 50 001 до 200 000           8.498210  
от 0 до 30 000                 9.090909

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [32]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Выясним распределение по целям кредита

data['purpose_category'].value_counts()

операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64

In [33]:
# Выясним влиянеие целей на возврат ккредита

purpose_real_estate = data[data['purpose_category'] == 'операции с недвижимостью']['total_income_category'].count()
purpose_real_estate_debt = data[(data['purpose_category'] == 'операции с недвижимостью') & (data['debt'] == 1)]['purpose_category'].count()
purpose_car = data[data['purpose_category'] == 'операции с автомобилем']['total_income_category'].count()
purpose_car_debt = data[(data['purpose_category'] == 'операции с автомобилем') & (data['debt'] == 1)]['purpose_category'].count()
purpose_education = data[data['purpose_category'] == 'получение образования']['total_income_category'].count()
purpose_education_debt = data[(data['purpose_category'] == 'получение образования') & (data['debt'] == 1)]['purpose_category'].count()
purpose_wedding = data[data['purpose_category'] == 'проведение свадьбы']['total_income_category'].count()
purpose_wedding_debt = data[(data['purpose_category'] == 'проведение свадьбы') & (data['debt'] == 1)]['purpose_category'].count()

conversion_purpose_al_estate = purpose_real_estate_debt / purpose_real_estate
conversion_purpose_car = purpose_car_debt / purpose_car
conversion_purpose_education = purpose_education_debt / purpose_education
conversion_purpose_wedding = purpose_wedding_debt / purpose_wedding

print(f'Конверсия для заемщиков с целью получения кредита на операции с недвижимостью равна: {conversion_purpose_al_estate:.3%}')
print(f'Конверсия для заемщиков с целью получения кредита на операции с автомобилем равна: {conversion_purpose_car:.3%}')
print(f'Конверсия для заемщиков с целью кредита получение образования равна: {conversion_purpose_education:.3%}')
print(f'Конверсия для заемщиков с целью получения кредита на проведение свадьбы равна: {conversion_purpose_wedding:.3%}')


Конверсия для заемщиков с целью получения кредита на операции с недвижимостью равна: 7.255%
Конверсия для заемщиков с целью получения кредита на операции с автомобилем равна: 9.348%
Конверсия для заемщиков с целью кредита получение образования равна: 9.253%
Конверсия для заемщиков с целью получения кредита на проведение свадьбы равна: 7.912%


<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера V1<a class="tocSkip"></h2>

    
<b>На доработку🤔:</b>
Напоминаю про построение сводных таблиц.
</div>

In [34]:
data_pivot_purpose = data.pivot_table(index=['purpose_category'], values=['debt'], aggfunc=['sum', 'count'])
data_pivot_purpose.columns=['Кол-во просрочек', 'Количество заемщиков']
data_pivot_purpose['Доля просрочек'] = data_pivot_purpose['Кол-во просрочек'] / data_pivot_purpose['Количество заемщиков'] *100
data_pivot_purpose = data_pivot_purpose.sort_values(by = 'Доля просрочек')
data_pivot_purpose

Кол-во просрочек  Количество заемщиков  \
purpose_category                                                   
операции с недвижимостью               780                 10751   
проведение свадьбы                     183                  2313   
получение образования                  369                  3988   
операции с автомобилем                 400                  4279   

                          Доля просрочек  
purpose_category                          
операции с недвижимостью        7.255139  
проведение свадьбы              7.911803  
получение образования           9.252758  
операции с автомобилем          9.347978

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Все верно построил, так держать!
</div>

**Вывод:** 
1. Выгоднее всего давать кредиты на недвижимость.
2. Выдача кредитов на автомобиль и образование сопряжена с более высоким риском.

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Все верно, молодец!
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Причины могут быть различные, как человеческий фактор, так и сбои в системе. Но в данном случае отсутствие данный в столбце days_employed вело за собой отсутствие данный в столбце total_income или же наоборот.

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Эти причины могут возникать в реальной жизни, согласен с тобой.
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
Если оставить пропуски, то могут получится некорректные данные. Если заполнить и среней статисчической, то выйдет таже ситуация, что и с пропусками, только тут влияют аномальные данные. Поэтому медианный данные менее всего искозят реальную картину.

<div class="alert alert-success">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Абсолютно верны ответ, молодец!
</div>

### Шаг 4: общий вывод.

Напишите ваш общий вывод.
Самый идеальный заемщик - это человек без детей, состоящий или состоявший в официальных отношениях, имеющий уровень дохода выше среднего и берущий деньги на недвижимость.

<div class="alert alert-warning">
<h2> Комментарий ревьюера V1<a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В итоговом выводе нужно достаточно подробно написать основные пункты выявленных зависимостей. Стоит доработать вывод с учетом моих комментариев по ходу работы.

Разброс просрочек по каждой из категорий примерно одинаковый.
Разница в абсолютных цифрах не такая большая. 

Детально по категориям:

Зависимость между семейным положением и возвратом кредита в срок

По наличию детей:
- Нет детей - 7.54%
- Есть дети - 9.24%
    - 1 ребенок - 9.24%
    - 2 ребенка - 9.46%
    - 3 ребенка - 8.18%
    - 4 ребенка - 9.76%
Абсолютная разница между минимальным и максимальным значением - 2.22%
Относительная разница между минимальным и максимальным значением - 22.75%

Зависимость между семейным положением и возвратом кредита в срок

По семейному положению:
- вдовец / вдова - 6.63%
- в разводе - 7.07%
- женат / замужем - 7.56%
- гражданский брак - 9.31%
- не женат / не замужем - 9.76%
Абсолютная разница между минимальным и максимальным значением - 3.13%
Относительная разница между минимальным и максимальным значением - 32.07%

Зависимость между уровнем дохода и возвратом кредита в срок

По уровню дохода
- От 30_001 до 50_000 - 6.02%
- От 200_000 до 1_000_000 - 7.06%
- От 1_000_001 и выше  - 8.00%
- От 50_000 до 200_000 - 8.50
- От 0 о 30_000 - 9.09%

Абсолютная разница между минимальным и максимальным значением - 3.07%
Относительная разница между минимальным и максимальным значением - 33.77%

Влияние цели кредита на вероятность возврата кредита в срок

По цели кредита
- недвижимость - 7.26%
- свадьба - 7.91%
- образование - 9.25%
- автомобиль - 9.35%
Абсолютная разница между минимальным и максимальным значением - 2.09%
Относительная разница между минимальным и максимальным значением - 22.35%

Из этого можно сделать вывод, что люди со сверхдоходами не так часто нуждаются в кредитах, а также и то, что люди с малым количеством дохода также опасаются их брать.

Несмотря на схожий в процентном отношении разброс значений, наибольшее влияние на вероятность просрочки показывает именно семейное положение. Во вторую очередь, следует обращать внимание на уровень дохода. Наименьшее влияние оказывают цели кредита и наличие детей.

<div class="alert alert-success">
<h2> Комментарий ревьюера V2<a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Отличный вывод, все четко, понятно и структурировано. Вывод в полной мере отражает существующие зависимости, и по нему можно принимать дальнейшие решения.
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проекту**

Спасибо за твой проект! Я уверен, что ты приложил все усилия, чтобы сделать этот проект. Это видно по твоей работе, а сформулированные выводы отражают, насколько успешно ты осваиваешь материал. 

**Отмечу отдельные положительные моменты проекта🙂**:
 
* Ты полностью следовал всем этапам выполнения задания и справился почти со всем. Хорошая работа!
* Отдельно выделю некоторые выводы, которые ты смог сформулировать абсолютно точно!
* У тебя опрятный и читаемый код. Возможно это звучит банально, однако при работе с чужим кодом это крайне важно)


**На что стоит обратить внимание🤓**:
    
* Важно доработать итоговый вывод.
* Нужно построить сводные таблицы во всех пунктах.
* Какая особенность данных не позволяет сходу дать ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?"
* Остальные рекомендации ты найдешь в моих комментариях!


**Желаю удачи и жду твой проект на повторное ревью!**😉

<div style="border:solid Chocolate 2px; padding: 40px">

Ты отлично справился с данной работой! Видно, как ты успешно осваиваешь материал, применяешь полученные знания на практике. Мне было очень приятно проверять твою работу, надеюсь, мои комментарии и советы были для тебя полезными.

**Желаю удачи в последующем обучении и исследованиях!**😉